torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -224,6 +241,25 @@ void RRefContext::delAllUsers(std::chrono::milliseconds timeoutMillis) {
     rref_ptr->tryDel();
   }
 
+  // If an rref in the owners_ map has never been forked, we will never get a
+  // corresponding message from the forking node(s) telling us to delete the
+  // RRef. Hence we delete the RRef here. This can occur when a remote call is
+  // sent to self and times out.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    std::vector<RRefId> unforkedOwners;
+    for (const auto& it : owners_) {
+      auto rrefId = it.first;
+      if (forks_.find(rrefId) == forks_.end()) {
+        unforkedOwners.push_back(rrefId);
+      }
+    }
+    for (auto& rrefId : unforkedOwners) {
+      LOG(INFO) << "Removing unforked OwnerRRef with RRefId: " << rrefId;
+      auto iter = owners_.find(rrefId);
+      owners_.erase(iter);
```

Comment:
What's our protocol for potentially deleting the intrusive_ptr<RRef> in owners_ with the lock held?
In `delForkOfOwner()` we maintain the intrusive pointer until after the lock is dropped, perhaps to avoid excess work with the lock held.

(There's a separate comment about whether an O(owners_) operation is needed here, but I guess for delAllUsers, there probably should be.

Reply:
@jjlilley , it seems like the only place where the `deletedRRef` is used is to explicitly acquire GIL and reset it if it holds a py object. I'm assuming this is to properly cleanup the python objects, so maybe we need to do something similar here.  @mrshenli Do you have any thoughts on this?

Reply:
The O(owners_) point is true, though this function only runs once during RPC shutdown (and also has a couple blocking waits and RPC calls), although, if we are concerned about the performance, there may be some things we can do to make this portion function non-blocking .